_posts/2020-06-25-taming-cljs-advanced-compilation.md

```diff
@@ -0,0 +1,172 @@
+enabled, the CLJS compiler will minify the result with Closure Compiler's Advanced Optimizations.
+
+Closure Compiler's Advanced Optimizations mode can cause some errors in your build that can be uncomfortable to
+track down if you are new to Closure Compiler.
+On top of that, help might be hard to find if you are trying to search the web with the error message because the function
+names are all mangled up.
+
+It is better to avoid these errors by keeping the Advanced Compilation mode in mind from the start of the project. 
+I'll describe some common methods to help you to avoid and resolve advanced compilation issues, and 
+show how to resolve the more uncommon issue that was described in the beginning and what is the reason behind it.
+
+## Avoiding issues with Advanced Compilation
+
+It is recommended to use `":optimizations :advanced"` CLJS compiler option for your production builds.
+You will greatly reduce your final build size by using more aggressive advanced transformations such as dead code 
+removal and aggressive renaming.  
+Read more about the advanced transformations at here: [Closure Compiler compilation levels](https://developers.google.com/closure/compiler/docs/compilation_levels#advanced_optimizations)
+
+Different build tools such as lein-cljsbuild and shadow-cljs might pass different default options to Closure Compiler.
+I won't cover the differences in this blog post, but instead, I will provide some general guidelines on how to avoid
+issues with advanced compilation in your project.
+
+### Keep your externs in a good shape
+Extern is a mechanism for declaring names that should not be munged (renamed) by Closure Compiler.
+When you are using external libraries, make sure that they come [bundled](http://cljsjs.github.io/)
+with externs if you are using a build tool like lein-cljsbuild. Or, provide externs files of your own if needed. 
+Otherwise, Closure Compiler will munge references to externally defined symbols unintentionally during advanced 
+compilation causing errors found only later when running the compiled code.
+
+To add externs of your own, use a compiler option for example: `":externs ["externs.js"]"` and provide an
+externs.js file in your working directory.
+
+Also, It is good to use `":infer-externs true"` compiler option. This option will enable generating externs automatically
+for JavaScript interop calls.
+
+It is worth to know, that Closure Compiler includes externs for stable JavaScript APIs, but newer features that have an
+experimental status might not be included yet. The experimental APIs change very quickly, so it makes no sense to include
+them in the Closure Compiler tool. So, if you are planning to use some experimental features, make sure 
+you add them to your own externs file.
+
+Read more about using external JavaScript libraries from [here](http://lukevanderhart.com/2011/09/30/using-javascript-and-clojurescript.html)
+
+### Splitting code might cause name collisions
```

Comment:
Could this have a short intro on code splitting, followed by the potential problems it causes, before proceeding to the solution?

Reply:
I didn't want to get too detailed here. However, I added a link to Code Splitting manual and changed the wording a bit.